samcli/lib/build/build_graph.py

```diff
@@ -115,54 +187,122 @@ def _read(self):
+
+        layer_build_definitions_table = document.get(BuildGraph.LAYER_BUILD_DEFINITIONS, [])
+        for layer_build_definition_key in layer_build_definitions_table:
+            layer_build_definition = _toml_table_to_layer_build_definition(layer_build_definition_key,
+                                                                     layer_build_definitions_table[
+                                                                         layer_build_definition_key])
+            self._layer_build_definitions.append(layer_build_definition)
+
+        # return self._build_definitions
 
     def _write(self):
         """
         Writes build definition details into build.toml file, which would be used by the next build.
         build.toml file will contain the same information as build graph,
         function details will only be preserved as function names
+        layer details will only be preserved as layer names
         """
         # convert build definition list into toml table
-        build_definitions_table = tomlkit.table()
-        for build_definition in self._build_definitions:
-            build_definition_as_table = _build_definition_to_toml_table(build_definition)
-            build_definitions_table.add(build_definition.uuid, build_definition_as_table)
+        function_build_definitions_table = tomlkit.table()
+        for build_definition in self._function_build_definitions:
+            build_definition_as_table = _function_build_definition_to_toml_table(build_definition)
+            function_build_definitions_table.add(build_definition.uuid, build_definition_as_table)
+
+        layer_build_definitions_table = tomlkit.table()
+        for build_definition in self._layer_build_definitions:
+            build_definition_as_table = _layer_build_definition_to_toml_table(build_definition)
+            layer_build_definitions_table.add(build_definition.uuid, build_definition_as_table)
 
         # create toml document and add build definitions
         document = tomlkit.document()
         document.add(tomlkit.comment("This file is auto generated by SAM CLI build command"))
-        document.add(BuildGraph.BUILD_DEFINITIONS, build_definitions_table)
+        document.add(BuildGraph.FUNCTION_BUILD_DEFINITIONS, function_build_definitions_table)
+        document.add(BuildGraph.LAYER_BUILD_DEFINITIONS, layer_build_definitions_table)
 
         if not self._filepath.exists():
             open(self._filepath, "a+").close()
 
         self._filepath.write_text(tomlkit.dumps(document))
 
 
-class BuildDefinition:
+class AbstractBuildDefinition:
     """
+    Abstract class for build definition
     Build definition holds information about each unique build
     """
 
-    def __init__(self, runtime, codeuri, metadata):
+    def __init__(self, source_md5):
+        self.uuid = str(uuid4())
+        self.source_md5 = source_md5
+
+    def set_source_md5(self, new_source_md5):
+        self.source_md5 = new_source_md5
```

Comment:
Do we need getters and setters instead of accessing the properties directly? Currently access patterns are inconsistent.

Reply:
Do you mean by using @property and @setter?

Reply:
In Python it's totally fine to do `my_object.attr`, i.e. without explicit getters/setters. It doesn't seem there's any need for them here as we're just assigning/retrieving values as-is, so my question is what is the instance attribute call pattern we're using in the codebase?

For example, I see both `build_definition.get_function_name()` and `build_definition.codeuri`, so which pattern are we using when?